### PR TITLE
Add notification policies and notification requests in web UI

### PR DESCRIPTION
**Pull request by Gargron**
Follow-up to #29366

____

Fixes MAS-39

<video src="https://github.com/mastodon/mastodon/assets/184731/34c312ac-28ee-47be-afb1-5adf8ae39211" />

### CODE REVIEW

File: app/javascript/mastodon/features/notifications/request.jsx [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+    element.focus();
+  }
+};
+
+export const NotificationRequest = ({ multiColumn, params: { id } }) => {
+  const columnRef = useRef();
+  const intl = useIntl();
+  const dispatch = useDispatch();
+  const notificationRequest = useSelector(state => state.getIn(['notificationRequests', 'current', 'item']));
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/notifications/request.jsx:46]: There might be a discrepancy between `notificationRequests.current` and the `id` param. This should probably be checked, with something like:
```suggestion
  const notificationRequest = useSelector(state => {
    const item = state.getIn(['notificationRequests', 'current', 'item']);
    return item?.get('id') === id : item : undefined;
  });
```

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
This doesn't constitute a full review, but I tried this on my server to help me review/evaluate the backend changes in #29366. I noticed a few bugs, issues, and suggestions:
- the requests' accounts are not properly imported—I made inline code suggestions to fix this
- once fetched, the list of potentially-filtered notifications does not change when you change a request, misattributing them—I made inline code suggestions for this too
- the list of notification requests appears to never get updated during the application's lifetime, which is likely going to be an issue for people who rarely if ever reload their app
- I think there is an opportunity to hide all media by default in the new filtered notification column

Otherwise, I really like how it looks!

**Comment by shleeable** [comment]
Those filter conditions are pretty good, but I'd be interested in bypassing the filter for "DM from somebody followed by somebody I follow".. 2 degrees would be safer from spammer

No idea about the performance overhead for that.

File: app/javascript/mastodon/actions/notifications.js [hunk 2]
```diff
@@ -313,3 +345,245 @@ export function setBrowserPermission (value) {
...
+    dispatch(fetchNotificationPolicySuccess(data));
+  }).catch(err => {
+    dispatch(fetchNotificationPolicyFail(err));
+  });
+};
+
+export const fetchNotificationRequests = () => (dispatch, getState) => {
+  if (!getState().getIn(['notificationRequests', 'items']).isEmpty() || getState().getIn(['notificationRequests', 'isLoading'])) {
+    return;
+  }
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/actions/notifications.js:386]: This prevents fetching newer (more recent) requests

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

**Review by ClearlyClaire — APPROVED** [review]
This looks good overall, but I think there are two issues left worth addressing quickly (either in this PR or follow-up PRs):
- unlike mutes, there is currently no interface to review and undo “dismissed” notification requests
- accepting a notification request does not immediately backfill the notifications column with the newly-accepted notifications

File: app/javascript/mastodon/locales/en.json [hunk 3]
```diff
@@ -271,6 +272,8 @@
   "filter_modal.select_filter.subtitle": "Use an existing category or create a new one",
   "filter_modal.select_filter.title": "Filter this post",
   "filter_modal.title.status": "Filter a post",
+  "filtered_notifications_banner.pending_requests": "Notifications from {count, plural, =0 {no} one {one person} other {# people}} you may know",
```
Comment (ciencia) [inline app/javascript/mastodon/locales/en.json:275]: when count is 0, source text says "Notifications from no you may know", which makes no sense

Comment (Gargron) [reply app/javascript/mastodon/locales/en.json:275]: That should've been "no one", but also, the banner is hidden when the count is 0.

**Pull request merged**